src/pymor/algorithms/samdp.py

```diff
@@ -0,0 +1,462 @@
+    function
+
+    .. math::
+        H(s) = C^T (s E - A)^{-1} B + D
+
+    of an LTI system. It is possible to take advantage of prior knowledge about the poles
+    by specifying shift parameters, which are injected after a new pole has been found.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    E
+        The |Operator| E or `None`.
+    B
+        The |Operator| B.
+    C
+        The |Operator| C.
+    nwanted
+        The number of dominant poles that should be computed.
+    init_shifts
+        A |NumPy array| containing shifts which are injected after a new pole has been found.
+    which
+        A string specifying the strategy by which the dominant poles and residues are selected.
+        Possible values are:
+            'LR':   Select poles with largest ||residual|| / ||Re(pole)||
+            'LS':   Select poles with largest ||residual|| / ||pole||
+            'LM':   Select poles with largest ||residual||
+
+    tol
+        Tolerance for the residual of the poles.
+    imagtol
+        Relative tolerance for imaginary parts of pairs of complex conjugate eigenvalues.
+    conjtol
+        Tolerance for the residual of the complex conjugate of a pole.
+    rqitol
+        If the residual is smaller than rqitol the two-sided Rayleigh quotient iteration
+        is executed.
+    maxrestart
+        The maximum number of restarts.
+    krestart
+        Maximum dimension of search space before performing a restart.
+    rqi_maxiter
+        Maximum number of iterations for the two-sided Rayleigh quotient iteration.
+
+    Returns
+    -------
+    poles
+        A |NumPy array| containing the computed dominant poles.
+    residues
+        A |NumPy array| containing the computed residues.
+    rightev
+        A |VectorArray| containing the right eigenvectors of the computed poles.
+    leftev
+        A |VectorArray| containing the left eigenvectors of the computed poles.
+    """
+
+    logger = getLogger('pymor.algorithms.samdp.samdp')
+
+    if E is None:
+        E = IdentityOperator(A.source)
+
+    B = B.as_range_array()
+    C = C.as_source_array()
+
+    B_defl = B.copy()
+    C_defl = C.copy()
+
+    k = 0
+    nrestart = 0
+    nr_converged = 0
+
+    X = A.source.empty()
+    Q = A.source.empty()
+    Qt = A.source.empty()
+    Qs = A.source.empty()
+    Qts = A.source.empty()
+    AX = A.source.empty()
+    V = A.source.empty()
+
+    H = np.empty((0, 1))
+    G = np.empty((0, 1))
+    poles = np.empty((1, 0))
+
+    if init_shifts is None:
+        st = 0.
+        shift_nr = 0
+        nr_shifts = 0
+    else:
+        st = init_shifts[0]
+        shift_nr = 1
+        nr_shifts = len(init_shifts)
+
+    shifts = init_shifts
+
+    while nrestart < maxrestart:
+        k = k + 1
+
+        sEmA = st * E - A
+        sEmAB = sEmA.apply_inverse(B_defl)
+        Hs = C_defl.dot(sEmAB)
+
+        y_all, _, u_all = spla.svd(Hs)  # use scipy.sparse.linalg.svds(Hs, k=1) instead?
```

Comment:
Since both `B` and `C` are converted to `VectorArrays`, this implicitly means that the number of inputs and outputs is small, so the dense SVD is fine.
The assumption about the number of inputs and outputs should be documented (or maybe `B` and `C` should be given as `VectorArrays`, as in `solve_lyap_lrcf`).